src/Storages/MergeTree/MergeTreeThreadSelectProcessor.cpp

```diff
@@ -39,28 +40,53 @@ MergeTreeThreadSelectProcessor::MergeTreeThreadSelectProcessor(
         min_marks_to_read = (min_marks_to_read_ * fixed_index_granularity + max_block_size_rows - 1)
             / max_block_size_rows * max_block_size_rows / fixed_index_granularity;
     }
+    else if (extension.has_value())
+    {
+        /// Parallel reading from replicas is enabled.
+        /// We try to estimate the average number of bytes in a granule
+        /// to make one request over the network per one gigabyte of data
+        /// Actually we will ask MergeTreeReadPool to provide us heavier tasks to read
+        /// because the most part of each task will be postponed
+        /// (due to using consistent hash for better cache affinity)
+        size_t sum_average_marks_size = 0;
+        auto column_sizes = storage.getColumnSizes();
+        for (const auto & name : extension->colums_to_read)
+        {
+            auto it = column_sizes.find(name);
+            if (it == column_sizes.end())
+                continue;
+            auto size = it->second;
+
+            if (size.data_compressed == 0 || size.data_uncompressed == 0 || size.marks == 0)
+                continue;
+
+            sum_average_marks_size += size.data_uncompressed / size.marks;
+        }
+
+        if (sum_average_marks_size == 0)
+            sum_average_marks_size = 8UL * 1024 * 1024 * 10; // 10Mib
+
+        min_marks_to_read = extension->count_participating_replicas * (/*1Gb*/ 8UL * 1024 * 1024 * 1024) / sum_average_marks_size;
```

Comment:
Why `8UL * 1024 * 1024 * 10` is 10Mib?

Reply:
And why do't we have a named constant yet ...

Reply:
...